File: mojang-rules/lib/ruleset.ts
import { z } from 'zod';
import { RuleSchema, type Rule, satisfiesRule } from './rule';
import type { OsOptions, OsName } from './os';

export type Ruleset = Rule[];

export const RulesetSchema: z.ZodType<Ruleset> = z.array(RuleSchema);

export function satisfiesRuleset(
  ruleset: Ruleset,
  os: OsOptions,
  feats: string[] = [],
): boolean {
  return ruleset.every((rule) => satisfiesRule(rule, os, feats));
}

export const emptyRuleset = (): Ruleset => [];

export const allowOsRuleset = (name: OsName): Ruleset => [
  { action: 'allow', os: { name } },
];

File: crates/torba-mojang-rules/src/ruleset.rs
use crate::os::{OsName, OsOptions, RuleError};
use crate::rule::{satisfies_rule, Rule, RuleAction};

pub type Ruleset = Vec<Rule>;

pub fn satisfies_ruleset(
    ruleset: &Ruleset,
    os: &OsOptions,
    feats: &[String],
) -> Result<bool, RuleError> {
    for rule in ruleset {
        if !satisfies_rule(rule, os, feats)? {
            return Ok(false);
        }
    }
    Ok(true)
}

pub fn empty_ruleset() -> Ruleset {
    Vec::new()
}

pub fn allow_os_ruleset(name: OsName) -> Ruleset {
    vec![Rule::Os {
        action: RuleAction::Allow,
        os: crate::os::OsConstraint {
            name: Some(name),
            ..Default::default()
        },
    }]
}
